server/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java

```diff
@@ -80,6 +89,24 @@ public TransportRolloverAction(
         this.rolloverService = rolloverService;
         this.client = client;
         this.activeShardsObserver = new ActiveShardsObserver(clusterService, threadPool);
+        this.rolloverTaskExecutor = (currentState, tasks) -> {
+            ClusterStateTaskExecutor.ClusterTasksResult.Builder<RolloverTask> builder = ClusterStateTaskExecutor.ClusterTasksResult
+                .builder();
+            ClusterState state = currentState;
+            for (RolloverTask task : tasks) {
+                try {
+                    state = task.performRollover(state);
```

Comment:
I do wonder, aren't we just using the `Metadata` here both as input and as we do the reroute later. Maybe we can get away with just loop over building `Metadata` which would be slightly cheaper? :)

Reply:
Eventually down in `MetadataCreateIndexService#clusterStateCreateIndex(...)` we do update the routing table (adding unassigned shards iirc). This gets invoked from the `MetadataRolloverService`. So I don't think this is possible?

Reply:
Yea this would require a little extra effort to do :) When I last profiled this it wasn't a big deal though, we can look into that kind of optimization later I supoose.